### PR TITLE
Fixed #31216 -- Added support for colorama terminal colors on Windows.

### PR DESCRIPTION
**Pull request by MinchinWeb**
If colorama is installed (as on Windows), support colored terminal output.

There is a very old ticket (https://code.djangoproject.com/ticket/13476) discussing how to support colorized terminal output on Windows. At the time, ANSICON was chosen as a solution over `colorama`. However, the link provided for ANSICON (http://adoxa.hostmyway.net/ansicon/) no longer works, and, personally, I struggled to install it previously (it involved messing around with DLL's and some fairly low level Windows functionality). `colorama`, on the other hand, is a current Python package that provides color on Windows by converting the ANSI cods to Windows native color "codes"; on other platforms, it just passes through the ANSI codes through. `colorama` is also very popular: for example, it is vendored by `pip` (https://github.com/pypa/pip/tree/master/src/pip/_vendor).

What this change does is test to see if `colorama` is available, and activate color (on Windows) if it is. It is not a hard dependency or even added to the "install_requires" of `setup.py`. Personally, I would support adding `colorama` as a dependency on Windows, but I wasn't sure how much support there would be for that.

No documentation changes have been added, although I think this would be wise to mention in the installation instructions for Windows. No tests have been added either, but I'm not really sure what to test directly. If someone could point me to where to make those changes, I would be happy to.

### CODE REVIEW

**Comment by MinchinWeb** [comment]
Ticket on TRAC -> https://code.djangoproject.com/ticket/31216

**Comment by MinchinWeb** [comment]
@carltongibson Thank you so much for the prompt comment! I will look into adding the documentation...

**Comment by MinchinWeb** [comment]
@carltongibson I've added documentation of this in four place:

* installation
* the Windows how-to
* django-admin
* version 3.1 release notes

The docs build for me locally, and look okay. Let me know your thoughts!

**Comment by MinchinWeb** [comment]
Question: How would I propose colorama being automatically installed (on Windows) with Django?

**Comment by carltongibson** [comment]
Hi @MinchinWeb. 

> Question: How would I propose colorama being automatically installed (on Windows) with Django?

What exactly would you propose? (We've tended to be skeptical about requiring dependencies in the past...)

**Comment by MinchinWeb** [comment]
> What exactly would you propose?

~~~yaml
# setup.cfg
install_requires =
    # existing
    asgiref >= 3.2
    pytz
    sqlparse >= 0.2.2
    # new
    colorama; platform_system=="Windows"
~~~

I'd leave the colorama check in ` django/core/management/color.py`. In this case, `colorama` would only be installed on Windows, but won't break anything if it's not present. I think it would go a long way to supporting Windows development.

File: docs/ref/django-admin.txt
Comment (carltongibson) [inline docs/ref/django-admin.txt]: `s/q/g/`

The created manage.py file has `if __name__ == '__main__'`, rather than a `manage()` function.

File: docs/ref/django-admin.txt
Comment (carltongibson) [inline docs/ref/django-admin.txt]: So, this trick works... 

Can we determine if colors are available? (Can we detect the new terminal?)

File: django/core/management/color.py
Comment (carltongibson) [inline django/core/management/color.py]: I think I'd have this as: 

```python
try:
  import colorama
  colorama.init()
except ImportError:
  colorama = None
```

Then below check `colorama is not None`.

File: docs/ref/django-admin.txt
Comment (carltongibson) [inline docs/ref/django-admin.txt]: I would drop this example. 

(Surely just `set ANSICON=on` is the easiest way to set an environment variable...)

**Comment by MinchinWeb** [comment]
> So per comment inline, **can we detect the new Windows Terminal?** That should be our first goal.

Maybe?? You can check for the environmental variable `WT_SESSION`, but that is also inherited by any child process you start from Windows Terminal (like if you start VS Code from the command line). See https://github.com/microsoft/terminal/issues/1040

Newer versions of PowerShell also support ANSI color codes, but I'm not sure which version it was added in, and detecting the that requires delving into the WinAPI.

**Comment by MinchinWeb** [comment]
I'm going to be away from my computer for the next three weeks, but I will follow up when I return!

**Comment by carltongibson** [comment]
Hi @MinchinWeb. Super thanks. 

I'll play some more with this but, I'm leaning towards thinking we should opt-in to the future here, perhaps just documenting an off-switch (if you're in an older environment). 

Let's discuss it and resolve when you're back.

File: docs/ref/django-admin.txt
Comment (MinchinWeb) [reply docs/ref/django-admin.txt:1816]: > (Surely just set ANSICON=on is the easiest way to set an environment variable...)

If only...

On Powershell `set ANSICON "on"` will indeed set a variable, but it is within the Powershell session, and not the environmental variable (so it's not going to be available to Django).

You could use `setx ANSICON "on"`, which will set the environmental variable, but this is actually adding an entry to the Registry, and so won't take effect until you re-open your terminal.

But as there seems to be support for installing colorama by default, I'll cut most of the section, including the example.

**Comment by MinchinWeb** [comment]
Hi @carltongibson ! I'm back!

I rebased the pull request onto the current master. It seems like a 3.1 branch has been created, so I updated the documentation to indicate this feature will be included in 3.2.

I like your suggestion to "opt-in to the future here", and so I've made installing `colorama` the default behaviour on Windows. This has the added advantage of being able to simplify much of the documentation of this feature; really I just note that this is a new feature, with a small note on the Windows page to turn it off using the `DJANGO_COLORS` environmental variable.

I also updated the "use color?" test like you'd suggested.

Let me know if you need me to do anything further. I'm excited to get this included in Django!

**Comment by MinchinWeb** [comment]
P.S. One thing I didn't do is deprecate ANSICON support, although I did remove its mention from the Window installation instructions.

File: setup.cfg
Comment (pope1ni) [inline setup.cfg:46]: Please keep this list sorted alphabetically.

File: docs/howto/windows.txt
Comment (pope1ni) [inline docs/howto/windows.txt:103]: `` :envvar:`DJANGO_COLORS` ``

**Comment by carltongibson** [comment]
Hi @MinchinWeb. Super. Welcome back. 😀 Thanks for picking this up.

File: setup.cfg
Comment (MinchinWeb) [reply setup.cfg:46]: done!

File: docs/howto/windows.txt
Comment (MinchinWeb) [reply docs/howto/windows.txt:103]: done!

File: docs/ref/django-admin.txt
Comment (MinchinWeb) [reply docs/ref/django-admin.txt]: I've dropped this example.

**Comment by MinchinWeb** [comment]
@carltongibson I'm excited to get this into Django. Is there anything holding this up? Is there anything I can do to move it forward? Thanks!

**Comment by carltongibson** [comment]
Hi @MinchinWeb — Yes, I'm excited too: I think this is a nice addition. 

>  Is there anything I can do to move it forward?

Not currently, I think. I just need the bandwidth to swing back to it. You're on my list. We'll get it in. Thanks for your patience.

File: django/core/management/color.py
Comment (carltongibson) [inline django/core/management/color.py:16]: Can we try moving this to a function, `has_colorama()` say? 

I know it's not much but the top-level import is a bit 🤔 — I think it might just be clearer inside a function. 
(maybe that looks horrible once we try it... — maybe then a comment would do...)

File: docs/howto/windows.txt
Comment (carltongibson) [inline docs/howto/windows.txt:105]: Add an instruction to `pip install colorama` here.

File: docs/ref/django-admin.txt
Comment (carltongibson) [inline docs/ref/django-admin.txt:1857]: I think we can drop these in favour of the install instructions (which link here.)

File: setup.cfg
Comment (carltongibson) [inline setup.cfg:44]: Can we drop this please? Ultimately we'd like to keep the list of `install_require` down to **required** packages.

File: django/core/management/color.py
Comment (MinchinWeb) [reply django/core/management/color.py:16]: If we move this to a function, we still need to have a global level variable to test to see if colorama is active (and thus we can turn on color, see line 24).

Also, I think moving this to a function would make it less clear what is going on. Also, this code really should only be called once (so it doesn't need to be in a function for DRY reasons). Possible replacement code:

```python

colorama = has_colorama()

def has_colorama():
     try:
        import colorama
    except ImportError:
        colorama = None
    else:
        colorama.init()

    return colorama
```

**Comment by MinchinWeb** [comment]
Hi @carltongibson ,

[Earlier](https://github.com/django/django/pull/12387#issuecomment-587367843), you'd said:

> I'll play some more with this but, I'm leaning towards thinking we should opt-in to the future here, perhaps just documenting an off-switch (if you're in an older environment).

`colorama` can be made an optional, manual installation step, but it's moving the other way. The change wouldn't be particularly hard: it would basically be a revert of 3c42c539bd350a27f71620dc658db06f66f885cb (an earlier commit to this pull request).

I also worry about discoverablity: i.e. the only people that will know about this are those that do a deep, deep dive into the documentation or they find it on a "Django tips and tricks" blog post.

If we do make `colorama` optional, do you want me to keep the `ANSICON` documentation?

If you still want it optional, let me know and I can make the changes needed.

**Comment by carltongibson** [comment]
Hi @MinchinWeb. Thanks. Yes. I was leaning that way. 

But after discussion with @felixxm we decided that including `colorama` in `install_requires` was a bit much. We've been cautious about taking on hard dependencies in the past, and for good reasons -- we'd like to maintain that. As such, we didn't feel we could justify the forced dependency. We'd prefer to leave it optional. I hope that makes sense. 

I guess leaving the old instructions don't hurt. I'd put the `colorama` install instructions first, since I think it's probably a better route for most people. 

Thanks for your efforts here!

**Comment by MinchinWeb** [comment]
Ok, just wanted to make sure something wasn't lost in the time between updates. I'll make the changes and push them here shortly!

**Comment by MinchinWeb** [comment]
I've made the changes you requested, and even figured out a reasonable way to make `colorama` NOT a top level import! I think everything is good to go; let me know if anything else is needed.

**Comment by MinchinWeb** [comment]
Hi @carltongibson !

I'm not quite out of energy yet on this, so I'll keep at it. But feel free to polish it off if need to hit a deadline.

I've adjusted the check, basically expanding `has_colorama()` into a `supported_platform()` function to allow for any number of checks; it was getting a little unwieldy as a one-liner. As well, this should support adding additional case fairly easy. The current checks are:

1. Not on Windows?
2. ANSICON?
3. Windows Terminal? (via the `WT_SESSION` environmental variable)
4. VS Code's Terminal? (via the 'TERM_PROGRAM` environmental variable)
4. Registry setting to support VT codes?
5. *colorama*?

It turns out my PowerShell terminal is like yours in that that it's not processing the color codes. I thought I'd "fixed" this some time ago, but maybe I'd just moved to Windows Terminal...

As for VS Code's terminal, I can't find it stated definitively either way whether it supports the VT codes/ANSI colors but it's working locally, so I'll go with it.

Not done here, but it seems you could enable ANSI colors using `SetConsoleMode` (part of the Win32 API). See https://github.com/ytdl-org/youtube-dl/issues/15758#issuecomment-370630896   This is beyond anything I've done in Python before, and am not sure I want to try to debug it when it breaks...

I haven't updated the documentation quite yet; but if the code works with you, I'll do that next.

**Comment by smithdc1** [comment]
`rich` has a [file](https://github.com/willmcgugan/rich/blob/master/rich/_windows.py) which prints the supported features of the terminal (vt & truecolor). 

Running this in powershell shows that it doesn't support `vt`, but the Windows terminal and the terminal in VSCode do.

**Comment by carltongibson** [comment]
Hi @MinchinWeb I read the code before the comment... - it doesn’t look like VS code is checked for, but you said you’d include it — I’ll look closer on Tuesday but, if it’s not there can we add it? Thanks.

**Comment by MinchinWeb** [comment]
...well then let me just the commit to support VS Code. It was sitting here on my computer, but didn't make it to GitHub earlier for some reason.

**Review by carltongibson — COMMENTED** [review]
Hi @MinchinWeb. 

Thanks again for this. I've rebased and squashed and adjusted the docs. (Please `git reset` to this point if you make any changes.) 

I'm just going to go over each version manually now. 

Then a last squish down, because I know @felixxm has comments, and I think we're good to go. 🥇

File: docs/ref/django-admin.txt
Comment (felixxm) [inline docs/ref/django-admin.txt:1891]: `:envvar:`

**Review by felixxm — COMMENTED** [review]
@MinchinWeb Thanks for this patch :+1: I left small comments.

File: django/core/management/color.py
Comment (pope1ni) [inline django/core/management/color.py:18]: Do we really need this inner function? We could just shortcut out early if not a TTY.

File: django/core/management/color.py
Comment (pope1ni) [reply django/core/management/color.py:29]: It is also "Windows Terminal", not "Microsoft Terminal".

File: django/core/management/color.py
Comment (pope1ni) [inline django/core/management/color.py:34]:
```suggestion
        if os.environ.get('TERM_PROGRAM') == 'vscode':
```

File: django/core/management/color.py
Comment (pope1ni) [inline django/core/management/color.py:38]: I'm not sure that it makes any sense checking the registry. There is no guarantee that you are using a terminal that uses this. So it could just result in a blanket "on" if this is set. It would be better to take the `ctypes` approach mentioned to check whether currently enabled for the actual terminal in use.

File: django/core/management/color.py
Comment (carltongibson) [reply django/core/management/color.py:60]: That's my fault: in an earlier version, the `colorama.init()` was at the module level, but I didn't think having that there for everyone was a good placing. 

> ...use it here and in `color_style()`

Makes sense.

File: docs/ref/django-admin.txt
Comment (carltongibson) [reply docs/ref/django-admin.txt:1891]: @felixxm Where's this linking to? 

```
WARNING: 'envvar' reference target not found: ANSICON="on"
```

**Comment by carltongibson** [comment]
Right, this works as expected as far as I can see. @MinchinWeb are you OK to reset to https://github.com/django/django/pull/12387/commits/49374b19404edfdb5dcadbf39e3f090fc24e726a and address @felixxm's comments?

**Review by pope1ni — CHANGES_REQUESTED** [review]
It feels rather verbose and unwieldy. Perhaps something like the following:

```python
try:
    import colorama
except ImportError:
    HAS_COLORAMA = False
else:
    HAS_COLORAMA = True
    colorama.init()

...

def supports_color():
    """
    Return True if the running terminal supports color, otherwise False.
    """
    def enabled_in_windows_registry():
        # Check the Windows Registry to see if VT code handling has been
        # enabled by default; see https://superuser.com/a/1300251/447564
        try:
            import winreg
        except ImportError:
            return False
        key = winreg.OpenKey(winreg.HKEY_CURRENT_USER, 'Console')
        try:
            value, _ = winreg.QueryValueEx(key, 'VirtualTerminalLevel')
        except FileNotFoundError:
            return False
        return value == 1

    # isatty() is not always implemented, #6223.
    return hasattr(sys.stdout, 'isatty') and sys.stdout.isatty() and (
        sys.platform != 'win32' or
        HAS_COLORAMA or
        'ANSICON' in os.environ or
        'WT_SESSION' in os.environ or  # Windows Terminal
        os.environ.get('TERM_PROGRAM') == 'vscode' or  # MVSC Terminal
        enabled_in_windows_registry()
    )
```

As mentioned below, I'd prefer that we switch out that inner function checking the registry for one that checks whether the terminal attached to `sys.stdout` supports VT codes.

File: docs/ref/django-admin.txt
Comment (pope1ni) [reply docs/ref/django-admin.txt:1891]: `:envvar:` won't work unless the docs have `.. envvar:: ANSICON` defined.

We've used it for those that are defined in Django and Python documentation, e.g. fbdb032de266ba5f82e061ab204f6c622889d563.

Comment (felixxm) [reply docs/ref/django-admin.txt:1891]: Right, my mistake.

File: django/core/management/color.py
Comment (carltongibson) [reply django/core/management/color.py:38]: I think the point is that this defaults to `off`, so our chance of a false positive is going to be pretty low no? (In which case folks can just set `nocolors`...) (But, yes, maybe this is a better way...)

The ctypes approach is from Davids https://github.com/django/django/pull/12387#issuecomment-674494510

**Comment by carltongibson** [comment]
We're going round in a circle on the `colorama.init()` call. https://github.com/django/django/pull/12387#discussion_r462840432

We should probably go back to the original (or Nick's variant). If we're going to have it at the top level, then we may as well init it there, to keep it in one place. Sorry for the back-and-forth on that @MinchinWeb. 

We should try and get this in one way or the other, as it's a big win for a lot of our users.

**Comment by smithdc1** [comment]
@MinchinWeb just to say thanks for all your work on this. I'm super excited for this feature.

File: django/core/management/color.py
Comment (MinchinWeb) [reply django/core/management/color.py:29]: Yes, "Windows Terminal" https://www.microsoft.com/en-us/p/windows-terminal/9n0dx20hk701?activetab=pivot:overviewtab

**Comment by carltongibson** [comment]
Hi @MinchinWeb. No stress! 

We're nearly there, and there's a long time to the 3.2 feature freeze. Plenty of time. 

If at any stage you think you'd rather pass it on, just say, and I'm happy to pick it up. 

Thanks again for your input!

**Comment by smithdc1** [comment]
Hi @MinchinWeb hope you are well, wow what a year it has been. 

We're a couple months out from the next feature freeze and I'm keen on seeing this get into 3.2.

Are you still up for finishing this off? Anything you need help with?

File: django/core/management/color.py
Comment (MinchinWeb) [reply django/core/management/color.py]: Fixed in 2ac0a53e18244cc6d1fbafe7cce967ace3e1fd65

File: django/core/management/color.py
Comment (MinchinWeb) [reply django/core/management/color.py]: I "trailing dots" I assumed you meant a period at the end of the sentence.

Fixed in 2ac0a53e18244cc6d1fbafe7cce967ace3e1fd65

File: django/core/management/color.py
Comment (MinchinWeb) [reply django/core/management/color.py]: Fixed in 2ac0a53e18244cc6d1fbafe7cce967ace3e1fd65

File: django/core/management/color.py
Comment (MinchinWeb) [reply django/core/management/color.py]: Fixed in 2ac0a53e18244cc6d1fbafe7cce967ace3e1fd65

File: docs/spelling_wordlist
Comment (MinchinWeb) [reply docs/spelling_wordlist:113]: colorama needs to be here or the Sphinx spell checker complains...

File: django/core/management/color.py
Comment (MinchinWeb) [reply django/core/management/color.py]: Resolved in 2ac0a53e18244cc6d1fbafe7cce967ace3e1fd65

File: docs/ref/django-admin.txt
Comment (MinchinWeb) [reply docs/ref/django-admin.txt:1896]: Which link in particular? All 3 of them?

File: docs/ref/django-admin.txt
Comment (MinchinWeb) [reply docs/ref/django-admin.txt]: Fixed in 2ac0a53e18244cc6d1fbafe7cce967ace3e1fd65

File: django/core/management/color.py
Comment (MinchinWeb) [reply django/core/management/color.py]: Fixed in 2ac0a53e18244cc6d1fbafe7cce967ace3e1fd65

File: docs/spelling_wordlist
Comment (smithdc1) [reply docs/spelling_wordlist:113]: You can add colorama in double backticks where it's mentioned instead of adding it here.

File: django/core/management/color.py
Comment (MinchinWeb) [reply django/core/management/color.py]: The setting defaults to "off" and so if it's on, you've most likely done it deliberately at some point. I.e. it seems like a deliberate attempt by the user to get color output.

It's possible that a terminal emulation might not pay attention to this setting, but that is only an issue if it also does not process VT colours. I'm not aware of any any that would fall into this (although I also haven't don't an exhaustive survey either...). I.e. This seems more hypothetical than something that we will run into with any frequency.

The `ctypes` approach is probably better, however, it is beyond my (current) Python ability. My fear there is that even if I did get it working, I wouldn't be able to fix it (debug it) when if breaks...

**Comment by MinchinWeb** [comment]
@carltongibson @smithdc1 I think this is ready to go.

I've added a comment to the testing documentation. I've tried to address the comments here. Let me know if anything else need polishing.

@smithdc1 Indeed, what a year! My year has been well, all things considered. Busy, very busy. We're happy and healthy, and last night's snowfall has left the world in beautiful white :)

File: django/core/management/color.py
Comment (MinchinWeb) [reply django/core/management/color.py]: Resolved in 2ac0a53e18244cc6d1fbafe7cce967ace3e1fd65

File: docs/spelling_wordlist
Comment (MinchinWeb) [reply docs/spelling_wordlist:113]: even if it's used as a link anchor text?

File: docs/ref/django-admin.txt
Comment (felixxm) [reply docs/ref/django-admin.txt:1896]: No, this one: `.. __: https://superuser.com/a/1300251/447564`.

**Review by carltongibson — APPROVED** [review]
OK, thanks @MinchinWeb, let's take this now. 

Zen of Python: 

> Now is better than never 

Maybe the ctypes approach is nicer overall, but this is a good win, seems robust enough in my testing, and is the option we have in hand now. If someone else wants to come forward later with a better patch then cool, but this works well enough. It's likely that's never needed. 

Thanks for your effort @MinchinWeb. Welcome aboard! ⛵️

**Comment by pope1ni** [comment]
> Maybe the ctypes approach is nicer overall, but this is a good win, seems robust enough in my testing, and is the option we have in hand now. If someone else wants to come forward later with a better patch then cool, but this works well enough. It's likely that's never needed.

:+1: I'd be able to handle the ctypes side of things, but I don't have access to a Windows environment so :man_shrugging:. My main interest is that Windows support is advanced enough not to hold back using colours or unicode on other platforms. This is nonetheless a great improvement. Thanks @MinchinWeb.

**Comment by felixxm** [comment]
Thanks y'all :confetti_ball: 

I pushed small edits.

**Comment by MinchinWeb** [comment]
@carltongibson @pope1ni @felixxm 

Thank you all 🙏 for all your help in getting this across the finish line! It's excited to say I helped make Django just a little bit better :) Thank you again!

**Comment by carltongibson** [comment]
Hi @MinchinWeb — Thank you for your work here. It's a great addition. Welcome aboard! ⛵️

**Pull request merged**
